unblob/handlers/archive/xiaomi/hdr1.py

```diff
@@ -0,0 +1,167 @@
+import binascii
+import io
+from pathlib import Path
+from typing import Optional
+
+import attrs
+from dissect.cstruct import Instance
+from structlog import get_logger
+
+from unblob.extractor import carve_chunk_to_file
+from unblob.file_utils import File, InvalidInputFormat, snull
+from unblob.models import (
+    Chunk,
+    Endian,
+    Extractor,
+    HexString,
+    StructHandler,
+    StructParser,
+    ValidChunk,
+)
+
+logger = get_logger()
+CRC_CONTENT_OFFSET = 12  # The CRC32 value is located after 12 byte in the header
+SIGNATURE_LEN = 272  # signature_header_t contains 4 bytes of size + 12 bytes for padding x3 + 0x100 is 256 in decimal
+BLOB_MAGIC = 0x000BEBA  # Blob header magic
+BLOB_HEADER_LEN = 48  # Blob header Length
+
+# https://lxr.openwrt.org/source/firmware-utils/src/xiaomifw.c
+C_DEFINITIONS = r"""
+    struct xiaomi_header {
+          char magic[4];                /* HDR1 */
+          uint32 signature_offset;
+          uint32 crc32;
+          uint16 unused;
+          uint16 device_id;             /* RA70 */
+          uint32 blob_offsets[8];
+    } hdr_header_t;
+
+    struct xiaomi_blob_header {
+          uint32 magic;                 /* 0x0000babe */
+          uint32 flash_offset;
+          uint32 size;                  /* Size of blob */
+          uint16 type;                  /* Type of blob */
+          uint16 unused;
+          char name[32];                /* Name of blob */
+    } blob_header_t;
+
+    struct xiaomi_signature_header {
+          uint32 size;
+          uint32 padding[3];
+          uint8 content[0x100];
+    } signature_header_t;
+ """
+
+
+@attrs.define
+class Blob:
+    offset: int
+    name: str
+    size: int
+
+
+class HDRExtractor(Extractor):
+    def extract(self, inpath: Path, outdir: Path):
+        with File.from_path(inpath) as file:
+            parser = HDRParser()
+            parser.parse(file)
+            for blob in parser.blobs:
+                chunk = Chunk(
+                    start_offset=blob.offset, end_offset=blob.offset + blob.size
+                )
+                carve_chunk_to_file(outdir.joinpath(Path(blob.name)), file, chunk)
+
+
+class HDRParser:
+    def __init__(self):
+        self._blobs = []
+        self._struct_parser = StructParser(C_DEFINITIONS)
+
+    def parse(self, file: File):
+        """Parse header + blob headers and populate self._blobs"""
+        header = self._struct_parser.parse("hdr_header_t", file, Endian.LITTLE)
+        for offset in header.blob_offsets:
+            if not offset:
+                continue
+
+            file.seek(offset, io.SEEK_SET)
+            blob_header = self._struct_parser.parse(
+                "blob_header_t", file, Endian.LITTLE
+            )
+
+            if not self.is_valid_header(blob_header, file):
+                raise InvalidInputFormat("Invalid blob header.")
+            file.seek(-BLOB_HEADER_LEN, io.SEEK_CUR)
```

Comment:
Remove this. If you keep it, the carved out file will contain both the blob and blob header. We only want the blob.